File: Medium/Java/TreeAndGraph/ZigzagLevelOrder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root != null) {
            boolean odd = true;
            List<TreeNode> level = new ArrayList<>();
            level.add(root);
            while (level.size() != 0) {
                List<Integer> l = new ArrayList<>();
                if (odd) {
                    odd = false;
                    for (int i = 0; i < level.size(); i ++)
                        l.add(level.get(i).val);
                }
                else {
                    odd = true;
                    for (int i = level.size() - 1; i >= 0; i --)
                        l.add(level.get(i).val);
                }
                res.add(new ArrayList(l));
                List<TreeNode> temp = new ArrayList<>();
                for (int i = 0; i < level.size(); i ++) {
                    temp.add(level.get(i).left);
                    temp.add(level.get(i).right);
                }
                level.clear();
                for (int i = 0; i < temp.size(); i ++)
                    if (temp.get(i) != null)
                        level.add(temp.get(i));
            }
        }
        return res;
    }
}

File: Medium/Python/TreesAndGraphs/ZigzagLevelOrder.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
            :type root: TreeNode
            :rtype: List[List[int]]
            """
        if not root:
            return []
        result, level = [], [root]
        reverse = False
        while level:
            if reverse:
                result.append([node.val for node in level[::-1]])
                reverse = False
            else:
                result.append([node.val for node in level])
                reverse = True
            temp = []
            for node in level:
                temp.extend([node.left, node.right])
            level = [leaf for leaf in temp if leaf]
        return result
